File: lib/kobako/outcome.rb
# frozen_string_literal: true

require_relative "codec"
require_relative "transport/error"

module Kobako
  # Host-facing boundary for the invocation outcome the native side split
  # off the core envelope. Takes the arm it named plus the fields that arm
  # carries, and settles the invocation the way a host does: return the
  # value, or raise the exception the failure attributes to.
  #
  # This is the two-step attribution decision. The wire framing belongs to
  # the native side; the payload codec at +Kobako::Codec+ decodes only
  # the one arm that carries a value.
  module Outcome
    # The two +origin+ values a Panic attributes with.
    ORIGIN_SANDBOX = "sandbox"
    ORIGIN_SERVICE = "service"

    # The guest-written class names that select a +SandboxError+ subclass.
    # A name absent here settles as plain +SandboxError+, so the guest
    # widens the taxonomy only by naming a class the host already defines.
    SUBCLASSES = {
      "Kobako::BytecodeError" => BytecodeError,
      "Kobako::UndefinedEntrypointError" => UndefinedEntrypointError
    }.freeze

    module_function

    # Settle one invocation. +kind+ names the arm, +payload+ is the
    # codec-encoded value the +:result+ arm carries, and +panic+ carries
    # the Panic's fields +[origin, class, message, backtrace, available]+ —
    # present on the panic arm and absent on every other, which is what
    # tells the failure that has a record to attribute from apart from the
    # two that do not. +entrypoint+ is the name this invocation asked for,
    # which the host knows and the wire therefore never carries.
    def reify(kind, payload, panic, entrypoint: nil)
      return decode_value(payload) if kind == :result

      raise panic ? panic_error(panic, entrypoint) : trap_error(kind)
    end

    # Map a Panic's fields onto the three-layer taxonomy. The fields land
    # on the exception verbatim — it carries the record rather than a
    # translation of one.
    def panic_error(panic, entrypoint)
      origin, klass, message, backtrace, available = panic
      selected = error_class(origin, klass)
      attribution = { origin: origin, klass: klass, backtrace_lines: backtrace }
      return selected.new(message, **attribution) unless selected == UndefinedEntrypointError

      UndefinedEntrypointError.new(message, name: entrypoint, available: available.map(&:to_sym), **attribution)
    end

    # +origin == "service"+ selects ServiceError; a sandbox-origin failure
    # naming one of the guest-written subclass names selects that subclass
    # so callers can rescue that path specifically.
    def error_class(origin, klass)
      return ServiceError if origin == ORIGIN_SERVICE

      SUBCLASSES.fetch(klass, SandboxError)
    end

    # An arm the host cannot settle: the guest wrote nothing, or wrote
    # bytes the envelope cannot frame. Either leaves nothing to attribute
    # to, so both walk the trap path; only the absent-versus-present
    # distinction selects the message.
    def trap_error(kind)
      return TrapError.new("Sandbox exited without producing a result") if kind == :absent

      TrapError.new(
        "Sandbox produced an unrecognised result; the runtime is corrupted, " \
        "discard this Sandbox before another invocation"
      )
    end

    # The Result arm's value — the one position a payload codec still
    # owns on this path. A decode fault means the framing was fine but the
    # carried value is unrepresentable.
    def decode_value(payload)
      # A Result is a payload position: an ext 0x02 Fault in it is a wire
      # violation, since a Fault's only home is a Reply's fault arm.
      Kobako::Codec.forbid_faults { Kobako::Codec::Decoder.decode(payload) }
    rescue Kobako::Codec::Error => e
      raise wire_error("Sandbox produced an invalid result value", diagnostic: e.message)
    end

    # Lift a wire violation the host detected to the real
    # +Kobako::Transport::Error+ class so callers can +rescue+ it
    # specifically instead of pattern-matching on +error.klass+. The
    # +klass+ field is still populated so existing operator-side tooling
    # that greps on the string continues to work.
    def wire_error(message, diagnostic: nil)
      Kobako::Transport::Error.new(
        message,
        origin: ORIGIN_SANDBOX,
        klass: "Kobako::Transport::Error",
        diagnostic: diagnostic
      )
    end

    # +reify+ is the whole seam: one call settles one invocation, matching
    # the single entry point the Rust frontend's twin exposes. Everything
    # else here is how that decision is reached.
    private_class_method :panic_error, :error_class, :trap_error, :decode_value, :wire_error
    private_constant :ORIGIN_SANDBOX, :ORIGIN_SERVICE, :SUBCLASSES
  end
end

File: crates/kobako/src/outcome.rs
//! Outcome-buffer classification: bytes → payload bytes or `Error`.
//!
//! The SDK twin of the Ruby gem's `Kobako::Outcome` module: split the
//! one-byte tag and attribute every failure the way the three-layer
//! taxonomy demands — so both frontends looking at the same outcome
//! bytes reach the same error variant. Attribution reads the envelope
//! alone; the Result arm's payload meets a schema later, on the
//! `Execution` accessor that asks for one. The wire-violation attribution string is the SPEC-pinned
//! wire-level error class name, not a Ruby leakage.

use kobako_transport::envelope::{Outcome, Panic};

use crate::error::{Error, Failure};

/// Classify one OUTCOME_BUFFER by its envelope alone: the Result arm's
/// payload bytes, or the `Error` its failure attributes to. Reads no
/// payload byte, so attribution works for a host whose Receivers speak
/// a schema this crate does not know.
pub(crate) fn classify(bytes: &[u8]) -> Result<Vec<u8>, Error> {
    match Outcome::decode(bytes) {
        Ok(Outcome::Result(payload)) => Ok(payload),
        Ok(Outcome::Panic(panic)) => Err(classify_panic(panic)),
        // Framing the outcome is the one thing the host does before
        // attribution, so a message it cannot frame — an empty buffer
        // included — leaves nothing to attribute to.
        Err(_) if bytes.is_empty() => Err(Error::Trap(
            "Sandbox exited without producing a result".into(),
        )),
        Err(_) => Err(Error::Trap(
            "Sandbox produced an unrecognised result; the runtime is corrupted, \
             discard this Sandbox before another invocation"
                .into(),
        )),
    }
}

/// `origin == "service"` → `Service`; a sandbox-origin panic carrying
/// the bytecode rejection class → `Bytecode`; everything else →
/// `Sandbox`. Every field is typed at the envelope, so classifying a
/// Panic reads no payload byte and cannot fail.
fn classify_panic(panic: Panic) -> Error {
    let from_service = panic.from_service();
    let failure = Box::new(Failure {
        class: panic.error.class,
        message: panic.error.message,
        backtrace: panic.error.backtrace,
        available: panic.available,
        diagnostic: None,
    });
    if from_service {
        Error::Service(failure)
    } else if failure.class == "Kobako::BytecodeError" {
        Error::Bytecode(failure)
    } else {
        Error::Sandbox(failure)
    }
}

#[cfg(test)]
mod tests {
    use kobako_transport::envelope::ErrorRecord;

    use super::*;

    fn panic_bytes(origin: &str, class: &str) -> Vec<u8> {
        Outcome::Panic(Panic {
            origin: origin.into(),
            error: ErrorRecord {
                class: class.into(),
                message: "boom".into(),
                backtrace: Vec::new(),
            },
            available: Vec::new(),
        })
        .encode()
    }

    #[test]
    fn the_result_arm_yields_the_payload_bytes_it_carried() {
        assert_eq!(
            classify(&Outcome::Result(vec![0x2a]).encode()).unwrap(),
            vec![0x2a],
            "a Result arm through classify must hand back its payload bytes untouched, \
             since attribution reads no payload byte"
        );
    }

    #[test]
    fn service_origin_panic_becomes_service_error() {
        let result = classify(&panic_bytes("service", "Kobako::ServiceError"));
        assert!(matches!(result, Err(Error::Service(f)) if f.message == "boom"));
    }

    #[test]
    fn bytecode_class_panic_becomes_bytecode_error() {
        let result = classify(&panic_bytes("sandbox", "Kobako::BytecodeError"));
        assert!(matches!(result, Err(Error::Bytecode(_))));
    }

    #[test]
    fn sandbox_origin_panic_becomes_sandbox_error() {
        let result = classify(&panic_bytes("sandbox", "RuntimeError"));
        assert!(matches!(result, Err(Error::Sandbox(f)) if f.class == "RuntimeError"));
    }

    #[test]
    fn empty_bytes_walk_the_trap_path() {
        assert!(matches!(classify(&[]), Err(Error::Trap(_))));
    }

    #[test]
    fn unknown_tag_walks_the_trap_path() {
        assert!(matches!(classify(&[0x7f, 0x2a]), Err(Error::Trap(_))));
    }

    #[test]
    fn a_panic_record_the_envelope_cannot_frame_walks_the_trap_path() {
        // The Panic arm followed by a truncated origin length prefix.
        let result = classify(&[0x02, 0x00, 0x00]);
        assert!(
            matches!(result, Err(Error::Trap(_))),
            "a Panic the envelope cannot frame leaves nothing to attribute to, got {result:?}"
        );
    }

    // E-27: an unresolved entrypoint reaches the caller with the names it
    // could have been, matching what the Ruby frontend exposes as
    // `#available` on its own subclass.
    #[test]
    fn an_unresolved_entrypoint_carries_the_names_it_could_have_been() {
        let bytes = Outcome::Panic(Panic {
            origin: "sandbox".into(),
            error: ErrorRecord {
                class: "Kobako::UndefinedEntrypointError".into(),
                message: "undefined entrypoint: Wrker".into(),
                backtrace: Vec::new(),
            },
            available: vec!["Worker".into(), "Helper".into()],
        })
        .encode();
        let result = classify(&bytes);
        assert!(
            matches!(result, Err(Error::Sandbox(ref f)) if f.available == ["Worker", "Helper"]),
            "an unresolved entrypoint must reach the caller with its correction, got {result:?}"
        );
    }
}
